controllers/componentdetectionquery_controller.go

```diff
@@ -109,28 +111,83 @@ func (r *ComponentDetectionQueryReconciler) Reconcile(ctx context.Context, req c
+				r.SetCompleteConditionAndUpdateCR(ctx, &componentDetectionQuery, err)
+				return ctrl.Result{}, nil
+			}
+
+			devfileBytes, dockerfileBytes = devfile.DownloadDevfileAndDockerfile(gitURL)
+		} else {
+			// Use SPI to retrieve the devfile from the private repository
+			// TODO - maysunfaisal also search for Dockerfile
+			devfileBytes, err = spi.DownloadDevfileUsingSPI(r.SPIClient, ctx, componentDetectionQuery.Namespace, source.URL, "main", "")
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Unable to curl for any known devfile locations from %s %v", source.URL, req.NamespacedName))
+			}
+		}
+
+		isDevfilePresent = len(devfileBytes) != 0
+		isDockerfilePresent = len(dockerfileBytes) != 0
+
+		if isDevfilePresent {
+			log.Info(fmt.Sprintf("Found a devfile, devfile to be analyzed to see if a Dockerfile is referenced %v", req.NamespacedName))
+			devfilesMap["./"] = devfileBytes
+		} else if isDockerfilePresent {
+			log.Info(fmt.Sprintf("Determined that this is a Dockerfile only component  %v", req.NamespacedName))
+			dockerfileContextMap["./"] = "./Dockerfile"
+		}
+
+		// Clone the repo if no dockerfile present
+		if !isDockerfilePresent {
+			log.Info(fmt.Sprintf("Unable to find devfile or Dockerfile under root directory, run Alizer to detect components... %v", req.NamespacedName))
+
+			clonePath, err = ioutils.CreateTempPath(componentDetectionQuery.Name, r.AppFS)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Unable to create a temp path %s for cloning %v", clonePath, req.NamespacedName))
+				r.SetCompleteConditionAndUpdateCR(ctx, &componentDetectionQuery, err)
+				return ctrl.Result{}, nil
+			}
+
+			err = util.CloneRepo(clonePath, source.URL, gitToken)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Unable to clone repo %s to path %s, exiting reconcile loop %v", source.URL, clonePath, req.NamespacedName))
+				r.SetCompleteConditionAndUpdateCR(ctx, &componentDetectionQuery, err)
+				return ctrl.Result{}, nil
+			}
+			log.Info(fmt.Sprintf("cloned from %s to path %s... %v", source.URL, clonePath, req.NamespacedName))
+			components, err := recognizer.DetectComponents(clonePath)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Unable to detect components using Alizer for repo %v, under path %v... %v ", source.URL, clonePath, req.NamespacedName))
+				r.SetCompleteConditionAndUpdateCR(ctx, &componentDetectionQuery, err)
+				return ctrl.Result{}, nil
+			}
+			log.Info(fmt.Sprintf("components detected %v... %v", components, req.NamespacedName))
+			// If no devfile and no dockerfile present in the root
+			// case 1: no components been detected by Alizer, might still has subfolders contains dockerfile. Need to scan repo
+			// case 2: more than 1 components been detected by Alizer, is certain a multi-component project. Need to scan repo
+			// case 3: one or more than 1 compinents been detected by Alizer, and the first one in the list is under sub-folder. Need to scan repo.
```

Comment:
the difference is, when there is only one component detected by Alizer, but the first component context is a subfolder. it might be a multi-component project, since other subfolders may contain dockerfile couldn't recognized by Alizer